File: Data Structures & Algorithms Module/Week7/SumDigits.java
/*
SumDigits.java
Billy Cussen
12/10/2020
*/

// Given a non-negative int n, return the sum of its digits recursively (no loops). Note
// that mod (%) by 10 yields the rightmost digit (126 % 10 is 6), while divide (/) by 10
// removes the rightmost digit (126 / 10 is 12).
public class SumDigits{   
   
    public static int sumDigits(int n) {
        int sum = 0, temp = 0;
        while(n > 0){
            temp = n % 10;
            sum += temp;
            n = n / 10;
        }
        return sum;
    }

    public static int sumDigitsRecursion(int n){
        if(n <= 0){
            return 0;
        }

        return (n % 10) + sumDigitsRecursion(n / 10);
    }

    public static void main(String [] args){
        long startTime = System.nanoTime();
        System.out.println("Looped Solution: "+sumDigits(124));
        long endTime = System.nanoTime();
        System.out.println("Looped Solution took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Recursive Solution: "+sumDigitsRecursion(124));
        endTime = System.nanoTime();
        System.out.println("Recursive Solution took "+(endTime-startTime)+" nanoseconds");
    }
}

File: Python/Data Structures & Algorithms Module - Python/Week7/SumDigits.py
"""
SumDigits.py
Billy Cussen
19/11/2020
"""

def sumDigits(a):
    sum = a
    while a > 0:
        a-=1
        sum+=a
    return sum

def sumDigitsRecurstion(a):
    if a == 0:
        return 0
    return a + sumDigitsRecurstion(a-1)


print("Sum Digits: "+str(sumDigits(5)))
print("Sum Digits Recursion: "+str(sumDigitsRecurstion(5)))